src/store/actions/auth.js

```diff
@@ -130,13 +149,94 @@ export const authGo = (email, password, isRegister) => {
   }
 };
 
+export const authSetProfileGo = (token, displayName, photoUrl, isEdit) => {
+  return (dispatch) => {
+    const apiKey = process.env.REACT_APP_FIREBASE_API_KEY;
+
+    const authData = {
+      idToken: token,
+      displayName: displayName,
+      photoUrl: photoUrl
+    }
+
+    let url = `https://identitytoolkit.googleapis.com/v1/accounts:update?key=${apiKey}`;
+
+    axios.post(url, authData)
+      .then((response) => {
+        dispatch(authDoneLoading());
+        dispatch(closeModal());
+        setTimeout(() => {
+          dispatch(openAndSetActiveModalAndMessage('success', `Your profile has been ${isEdit ? 'updated' : 'set'}!`));
+        }, 250);
+        dispatch(authGetProfile(token));
+      }).catch((error) => {
+        const errorMessage = setErrorMessage(error);
+
+        dispatch(authFail(errorMessage));
+        dispatch(closeModal());
+        setTimeout(() => {
+          dispatch(openAndSetActiveModalAndMessage('error', errorMessage));
+        }, 250);
+      });
+  }
+}
+
+export const authSetProfile = (token, displayName, photoUrl, isEdit) => {
```

Comment:
By taking the contents of conditional blocks and moving them out to functions, you increase your readability/extensibility/debuggability score:

```
const handlePhotoUpdated = ({ dispatch, token, displayName, photoUrl, isEdit }) => {
  const ref = firebaseStorageRef.child(photoUrl.name);

  ref.put(photoUrl)
    .then((response) => {
      return ref.getDownloadURL();
    })
    .then((responsePhotoUrl) => {
      dispatch(authSetProfileGo(token, displayName, responsePhotoUrl, isEdit));
    });
}

export const authSetProfile = (token, displayName, photoUrl, isEdit) => {
  return (dispatch) => {
    dispatch(authStart());

    const photoUpdated = typeof(photoUrl) === 'object';
    if(photoUpdated) {
      handlePhotoUpdated({ dispatch, token, displayName, photoUrl, isEdit });
    } else {
      dispatch(authSetProfileGo(token, displayName, photoUrl, isEdit));
    }
  }
}
```

Also, you expose what looks like "complex" logic down to simple logic (e.g. "Oh look! A ternary!"):

```
export const authSetProfile = (token, displayName, photoUrl, isEdit) => {
  return (dispatch) => {
    dispatch(authStart());

    const photoUpdated = typeof(photoUrl) === 'object';
    photoUpdated ?
      handlePhotoUpdated({ dispatch, token, displayName, photoUrl, isEdit }) :
      dispatch(authSetProfileGo(token, displayName, photoUrl, isEdit));
  }
}
```

Additionally, I'd recommend object arguments for functions instead of long-running lists in most cases (the saying goes that the best number of arguments for a JS function is 1):

```
// 'authProfileArgs' has a shape of '{ token, displayName, photoUrl, isEdit }'
export const authSetProfile = (authProfileArgs) => {
	return (dispatch) => {
		dispatch(authStart());

		const photoUpdated = typeof(authProfileArgs.photoUrl) === 'object';
		photoUpdated ?
			handlePhotoUpdated({ dispatch, ...authProfileArgs }) :
			dispatch(authSetProfileGo(authProfileArgs); // this is assuming that 'authSetProfileGo' has changed its function signature as well
	}
}```

Reply:
@shoelessness I guess this is similar to the previous one, but I get that it might be better here because you can then use a ternary.

For the second part, this makes sense but I'm wondering how it could really make much of a difference if you're making an object with all of the same items as you would put separately. Is it just a best practice or am I missing something maybe?

Reply:
By making an arguments list an arguments object, or multiple objects, you get the benefits of being able to pass the entire object without having to name each property, as well as being able to perform operations on the object without having to compose it somehow.

The other thing is that you can still break common sets of args out to separate object arguments. Say you have a map location lookup somewhere which requires the following:

```
getMapResponse(userName, userAuthToken, mapLat, mapLng, mapZoom);
```

You _could_ put all of them in one object, or you could split it up into two with separate concerns:

```
const userDetails = { userName, userAuthToken };
const mapDetails = { mapLat, mapLng, mapZoom);
getMapResponse(userDetails, mapDetails);
```

Inside of `getMapResponse`, say you had two helper functions; one for authenticating the user and one for doing the map lookup. In the first version with the flat list of arguments you'd have to do this:

```
const getMapResponse = (userName, userAuthToken, mapLat, mapLng, mapZoom) => {
  authenticateUser(userName, userAuthToken);
  lookupMap(mapLat, mapLng, mapZoom);
}
```

Whereas in the reworked version you get:

```
const getMapResponse = (userDetails, mapDetails) => {
  authenticateUser(userDetails);
  lookupMap(mapDetails);
}
```

Notice how the function signatures of `authenticateUser` and `lookupMap` would have to change. For example, authenticateUser would go from:

```
const authenticateUser = (userName, userAuthToken) => {
  // do the auth call
}
```

to (with object destructuring implemented):

```
const authenticateUser = ({ userName, userAuthToken }) => {
  // do the auth call
}
```

This actually makes your code more simple to read and reason about, which means easier extension and fewer bugs.

Reply:
@shoelessness So you mean that the benefit is that you don't have to include all of the properties it calls for? Also, what do you mean by "as well as being able to perform operations on the object without having to compose it somehow."?

The part about splitting into separate objects definitely makes sense though if you're doing the actions functional like you suggested.

Reply:
It means you can just pass around one object arg instead of a list of arguments which could be of an arbitrary length. Sure, one or two or three args makes sense, but when you wind up having like 12 in a row, all of a sudden it's unwieldy to read and to rewrite. Think of the way React handles this. You just get `props`, which is always clear what it means.

As for operating, when you have function arguments, you [shouldn't modify them](https://eslint.org/docs/rules/no-param-reassign). Having an object argument means you can do something like an object spread or Object.assign() to use them. Alternatively, if you have an argument which is an array, you can use array methods. If you just have the flat list, you have to reconstruct the list into some other data structure _before_ you can perform an operation.